client/client_update.go

```diff
@@ -191,7 +191,35 @@ func (u *UpdateClient) FetchUpdate(api ApiRequester, url string, maxWait time.Du
 
 type UpdateResponse struct {
 	*datastore.UpdateInfo
-	*updatecontrolmap.UpdateControlMap `json:"update_control_map"`
+
+	// Use a strict parser for the Update Control Map in order to reject
+	// unknown fields. This is in contrast to the rest of the response,
+	// where we allow unknown fields.
+	UpdateControlMap strictUpdateControlMapParser `json:"update_control_map"`
+}
+
+type strictUpdateControlMapParser struct {
+	*updatecontrolmap.UpdateControlMap
+
+	// Used to ignore the map after an error has been returned. We need to
+	// reparse without errors then, to get the UpdateInfo.
+	ignoreMap bool
+}
+
+func (s *strictUpdateControlMapParser) UnmarshalJSON(data []byte) error {
```

Comment:
How about we add the marshaller to the `UpdateInfo` type, and then the `disallowUnknownFields` to the udpate control map, and we can do this in one fell swoop?

Reply:
But what's the benefit? `UpdateInfo` only uses defaults, so it doesn't need an unmarshaller. Maybe I'm misunderstanding...

Reply:
I'm sorry, I meant the `UpdateResponse` struct of course!

Reply:
This doesn't work, because `DisallowUnknownFields` is a property of the parser, so you can't use it on only one sub element. The sub element itself needs to have an unmarshaller which sets it. It would be nice if it was settable with a JSON tag like `omitempty`, but I didn't find anything like that.

Reply:
Oh, and just to be clear on this point as well: We can't do two parse passes, one over `UpdateInfo` and one over `UpdateControlMap`, because the latter's key is not available before you have started parsing, it's just a blob of bytes at this point.

Reply:
Ur, right. thx